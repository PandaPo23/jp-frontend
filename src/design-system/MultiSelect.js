import styled from '@emotion/styled';
import PropTypes from 'prop-types';
import React from 'react';
import { components } from 'react-select';
import CreatableSelect from 'react-select/lib/Creatable';

import { selectOptionType } from '../utils/propTypes';
import Box from './Box';
import IconButton from './IconButton';
import Label from './Label';
import theme from './theme';

/**
 *
 * Example Usage
 *
 * <MultiSelect
 *   options={app.optionsFor('activities')}
 *   selected={app.selectedOptionsFor('activities')}
 *   label="Activities..."
 *   direction="ltr"
 *   onSelect={(selected) => app.selectOptionsFor('activities', selected)}
 *   onUnselect={(value) => app.unselectOptionsFor('activities', value)}
 *   onAdd={(input)=> app.addOptionFor('activities', input)}
 *   onSearch={(input) => app.searchOptionFor('activities', input)}
 * />
 *
 **/

/**************************************
 * custom renderer of selected values
 * difficult to customize react-select
 * to match design
 ***************************************/
const SelectedTagWrapper = styled('div')`
  display: flex;
  align-items: center;
  border: 1px solid ${theme.colors.selects.tagBorder};
  margin: 0.25rem;
  padding: 0 0.25rem;
`;

const SelectedTagsContainer = styled('div')`
  display: flex;
  flex-wrap: wrap;
  margin-top: 0.5rem;
  margin-left: -0.25rem;
`;

function SelectedTag({ data: { value, label }, onRemove }) {
  const handleRemove = () => {
    onRemove(value);
  };

  return (
    <SelectedTagWrapper>
      <IconButton
        size={14}
        name="clear"
        p={1}
        onClick={handleRemove}
        color={theme.colors.on.surface[0]}
      />
      <Label fontSize={4} fontWeight="normal" color="on.surface" p={1}>
        {label}
      </Label>
    </SelectedTagWrapper>
  );
}

/*************************************
 * customized styles for react-select
 **************************************/
const stylesForSelect = {
  placeholder: (styles, { isFocused }) => ({
    ...styles,
    color: isFocused ? theme.colors.on.surface[2] : theme.colors.on.surface[1],
  }),
  option: (styles, { isFocused }) => ({
    ...styles,
    cursor: 'pointer',
    padding: '0.5rem 1rem',
    backgroundColor: isFocused
      ? theme.colors.selects.optionFocusBg
      : theme.colors.surface[1],
    color: isFocused
      ? theme.colors.selects.primaryColor
      : theme.colors.on.surface[1],
  }),
  menu: (styles) => ({
    ...styles,
    marginTop: '2px',
    border: `1px solid ${theme.colors.primaryGray}`,
    borderRadius: 0,
    boxShadow: `0 1px 3px 0 ${theme.colors.selects.menuShadow}`,
  }),
  valueContainer: (styles) => ({
    ...styles,
    paddingLeft: 0,
    paddingRight: 0,
  }),
  control: (styles, { isFocused, menuIsOpen }) => ({
    ...styles,
    border: 'none',
    cursor: isFocused ? 'text' : 'pointer',
    borderRadius: 0,
    boxShadow: menuIsOpen
      ? `0 2px 0 0 ${theme.colors.selects.primaryColor}`
      : isFocused
      ? `0 1px 0 0 ${theme.colors.selects.primaryColor}`
      : `0 1px 0 0 ${theme.colors.primaryGray}`,
  }),
  dropdownIndicator: (styles, data) => ({
    ...styles,
    padding: 0,
    color: data.selectProps.menuIsOpen
      ? theme.colors.on.surface[2]
      : theme.colors.primaryGray,
  }),
  noOptionsMessage: (styles) => ({
    ...styles,
    color: theme.colors.on.surface[1],
  }),
};

const DropdownIndicator = (props) => (
  <components.DropdownIndicator {...props}>
    <IconButton
      size={24}
      name={props.selectProps.menuIsOpen ? 'arrow-drop-up' : 'arrow-drop-down'}
      px={0}
    />
  </components.DropdownIndicator>
);

/***************************************
 * main component rendering react-select
 ****************************************/
function ReactSelectWrapper({
  selected,
  label,
  direction,
  onSelect,
  onUnselect,
  onAdd,
  onSearch,
  ...rest
}) {
  const handleInputChange = (value, { action }) => {
    if (action === 'input-change') {
      onSearch(value);
    }
  };

  return (
    <Box>
      <CreatableSelect
        {...rest}
        placeholder={label}
        value={selected}
        isRtl={direction !== 'ltr'}
        onChange={onSelect}
        onCreateOption={onAdd}
        onInputChange={handleInputChange}
        isMulti
        styles={stylesForSelect}
        controlShouldRenderValue={false}
        backspaceRemovesValue={false}
        components={{
          ClearIndicator: null,
          IndicatorSeparator: null,
          DropdownIndicator,
        }}
      />

      <SelectedTagsContainer>
        {selected.map((tag) => (
          <SelectedTag key={tag.value} data={tag} onRemove={onUnselect} />
        ))}
      </SelectedTagsContainer>
    </Box>
  );
}

// const MultiSelect = styled(ReactSelectWrapper)``;
const MultiSelect = ReactSelectWrapper;

MultiSelect.propTypes = {
  options: PropTypes.arrayOf(selectOptionType).isRequired,
  selected: PropTypes.arrayOf(selectOptionType).isRequired,
  label: PropTypes.string,
  direction: PropTypes.string,
  onSelect: PropTypes.func.isRequired,
  onAdd: PropTypes.func,
  onSearch: PropTypes.func,
};

MultiSelect.defaultProps = {
  label: 'Select...',
  direction: 'ltr',
};

export default MultiSelect;
