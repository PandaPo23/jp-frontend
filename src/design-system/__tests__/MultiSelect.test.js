import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import MultiSelect from '../MultiSelect';

const options = [
  { value: 'fishing', label: 'Fishing' },
  { value: 'scuba', label: 'Scuba' },
  { value: 'climbing', label: 'Climbing' },
  { value: 'hunting', label: 'Hunting' },
];

describe('MultiSelect component', () => {
  it('renders initial state correctly', () => {
    const localProps = {
      options,
      selected: [],
      label: 'Activities...',
      onSelect: () => {},
      onUnselect: () => {},
      onAdd: () => {},
      onSearch: () => {},
    };
    const wrapper = shallow(<MultiSelect {...localProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders added items correctly', () => {
    const localProps = {
      options,
      selected: [options[1], options[2]],
      label: 'Activities...',
      onSelect: () => {},
      onUnselect: () => {},
      onAdd: () => {},
      onSearch: () => {},
    };
    const wrapper = shallow(<MultiSelect {...localProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
