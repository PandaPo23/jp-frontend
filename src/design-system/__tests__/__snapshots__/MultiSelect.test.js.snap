// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect component renders added items correctly 1`] = `
<Box>
  <StateManager
    backspaceRemovesValue={false}
    components={
      Object {
        "ClearIndicator": null,
        "DropdownIndicator": [Function],
        "IndicatorSeparator": null,
      }
    }
    controlShouldRenderValue={false}
    defaultInputValue=""
    defaultMenuIsOpen={false}
    defaultValue={null}
    isMulti={true}
    isRtl={false}
    onChange={[Function]}
    onCreateOption={[Function]}
    onInputChange={[Function]}
    options={
      Array [
        Object {
          "label": "Fishing",
          "value": "fishing",
        },
        Object {
          "label": "Scuba",
          "value": "scuba",
        },
        Object {
          "label": "Climbing",
          "value": "climbing",
        },
        Object {
          "label": "Hunting",
          "value": "hunting",
        },
      ]
    }
    placeholder="Activities..."
    styles={
      Object {
        "control": [Function],
        "dropdownIndicator": [Function],
        "menu": [Function],
        "noOptionsMessage": [Function],
        "option": [Function],
        "placeholder": [Function],
        "valueContainer": [Function],
      }
    }
    value={
      Array [
        Object {
          "label": "Scuba",
          "value": "scuba",
        },
        Object {
          "label": "Climbing",
          "value": "climbing",
        },
      ]
    }
  />
  <Styled(div)>
    <SelectedTag
      data={
        Object {
          "label": "Scuba",
          "value": "scuba",
        }
      }
      key="scuba"
      onRemove={[Function]}
    />
    <SelectedTag
      data={
        Object {
          "label": "Climbing",
          "value": "climbing",
        }
      }
      key="climbing"
      onRemove={[Function]}
    />
  </Styled(div)>
</Box>
`;

exports[`MultiSelect component renders initial state correctly 1`] = `
<Box>
  <StateManager
    backspaceRemovesValue={false}
    components={
      Object {
        "ClearIndicator": null,
        "DropdownIndicator": [Function],
        "IndicatorSeparator": null,
      }
    }
    controlShouldRenderValue={false}
    defaultInputValue=""
    defaultMenuIsOpen={false}
    defaultValue={null}
    isMulti={true}
    isRtl={false}
    onChange={[Function]}
    onCreateOption={[Function]}
    onInputChange={[Function]}
    options={
      Array [
        Object {
          "label": "Fishing",
          "value": "fishing",
        },
        Object {
          "label": "Scuba",
          "value": "scuba",
        },
        Object {
          "label": "Climbing",
          "value": "climbing",
        },
        Object {
          "label": "Hunting",
          "value": "hunting",
        },
      ]
    }
    placeholder="Activities..."
    styles={
      Object {
        "control": [Function],
        "dropdownIndicator": [Function],
        "menu": [Function],
        "noOptionsMessage": [Function],
        "option": [Function],
        "placeholder": [Function],
        "valueContainer": [Function],
      }
    }
    value={Array []}
  />
  <Styled(div) />
</Box>
`;
